elastic/kibana
# [Infra] Deprecate fields returned by the `/api/metrics/source` endpoint, in favor of ad hoc data views

closes [#180689](https://github.com/elastic/kibana/issues/180689)

## Summary

This PR lays the necessary foundation to deprecate the indices [fields list attribute retrieved and returned](https://github.com/elastic/kibana/blob/main/x-pack/plugins/observability_solution/infra/server/routes/metrics_sources/index.ts#L38) by the `api/metrics/source` endpoint and affects all infra pages.

Now instead of the Metrics UI and Inventory UI relying on that attribute to build the list of suggestions in their search bars, an adhoc data view will be created from the `metricsAlias` returned by the same endpoint. The data view object contains everything that's needed with the benefit of allowing us to make async requests and cache, provided by Data Views Service.

This will allow us to start moving away from `api/metrics/source` to retrieve information critical to the loading of the pages.

It's a big PR, and the changes affect mostly the search bars, field name selectors, and charts across infra. Besides, it also contains 
- Storybook fix.
- Some components were turned into functional components
- `<MetricsPageTemplate />` now encapsulates error messages and is only used in Infrastructure pages
- Improvement of the Infrastructure routing code.
- Data View provider common to all infra pages
- `SourceProvider` refactor.

<img width="921" alt="image" src="https://github.com/elastic/kibana/assets/2767137/ceb781ed-2464-4f67-93e4-7013786830d4">

<img width="921" alt="image" src="https://github.com/elastic/kibana/assets/2767137/169f6f6b-03da-48e3-86b8-449bad1b2035">

<img width="921" alt="image" src="https://github.com/elastic/kibana/assets/2767137/3e5b8394-ead2-40bb-ad00-609b29b179e5">

<img width="921" alt="image" src="https://github.com/elastic/kibana/assets/2767137/34639ebf-af75-46e3-840b-8aa89951ae6e">

<img width="921" alt="image" src="https://github.com/elastic/kibana/assets/2767137/b6ff3b96-219b-4daa-bc33-f9fd36912a5d">

<img width="921" alt="image" src="https://github.com/elastic/kibana/assets/2767137/437cb7b8-62a6-4e1a-b94e-41055ab71bc1">

<img width="921" alt="image" src="https://github.com/elastic/kibana/assets/2767137/72f04712-7ee1-44fc-990e-83f0fbff7115">

## How to test

- Start a local kibana instance pointing to an Oblt cluster
- Navigate to Inventory, and all its pages, check if charts, search bars and field name autocomplete fields are working
